diff -uNr a/source/common/engine/i_interface.cpp b/source/common/engine/i_interface.cpp
--- a/source/common/engine/i_interface.cpp	2023-02-28 12:20:33.000000000 +0100
+++ b/source/common/engine/i_interface.cpp	2023-02-28 17:10:07.028061190 +0100
@@ -5,7 +5,6 @@
 #include "c_cvars.h"
 #include "gstrings.h"
 
-static_assert(sizeof(void*) == 8, "32 builds are not supported");
 
 // Some global engine variables taken out of the backend code.
 FStartupScreen* StartWindow;
diff -uNr a/source/common/rendering/gles/gles_system.cpp b/source/common/rendering/gles/gles_system.cpp
--- a/source/common/rendering/gles/gles_system.cpp	2023-02-28 12:20:34.000000000 +0100
+++ b/source/common/rendering/gles/gles_system.cpp	2023-03-01 08:06:11.221431467 +0100
@@ -11,7 +11,7 @@
 EXTERN_CVAR(Bool, gl_customshader);
 void setGlVersion(double glv);
 
-
+#define __ANDROID__
 #if USE_GLAD_LOADER
 
 PFNGLMAPBUFFERRANGEEXTPROC glMapBufferRange = NULL;
@@ -206,6 +206,7 @@
 			gles.shaderVersionString = "100";
 			gles.depthStencilAvailable = CheckExtension("GL_OES_packed_depth_stencil");
 			gles.npotAvailable = CheckExtension("GL_OES_texture_npot");
+			gles.useMappedBuffers = true;
 			gles.depthClampAvailable = CheckExtension("GL_EXT_depth_clamp");
 			gles.anistropicFilterAvailable = CheckExtension("GL_EXT_texture_filter_anisotropic");
 		}
diff -uNr a/source/common/rendering/gles/gles_system.h b/source/common/rendering/gles/gles_system.h
--- a/source/common/rendering/gles/gles_system.h	2023-02-28 12:20:34.000000000 +0100
+++ b/source/common/rendering/gles/gles_system.h	2023-03-01 08:00:38.906856577 +0100
@@ -23,7 +23,7 @@
 #include <sys/stat.h>
 #include <fcntl.h>
 
-#define USE_GLAD_LOADER 0 // Set to 1 to use the GLAD loader, otherwise use noramal GZDoom loader for PC
+#define USE_GLAD_LOADER 1 // Set to 1 to use the GLAD loader, otherwise use noramal GZDoom loader for PC
 
 #if (USE_GLAD_LOADER)
 	#include "glad/glad.h"
diff -uNr a/source/games/duke/src/game_misc.cpp b/source/games/duke/src/game_misc.cpp
--- a/source/games/duke/src/game_misc.cpp	2023-02-28 12:20:34.000000000 +0100
+++ b/source/games/duke/src/game_misc.cpp	2023-03-01 10:34:01.645924666 +0100
@@ -236,8 +236,6 @@
 	DAngle cang;
 
 	pp = &ps[screenpeek];
-	// set palette here, in case the 3D view is off.
-	setgamepalette(setpal(pp));
 
 	float blend[4] = {};
 
