diff --git a/xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererDRMPRIMEGLES.cpp b/xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererDRMPRIMEGLES.cpp
index 18e1c882a261e..ae178ded44b21 100644
--- a/xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererDRMPRIMEGLES.cpp
+++ b/xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererDRMPRIMEGLES.cpp
@@ -30,6 +30,13 @@ using namespace KODI::UTILS::EGL;
 CRendererDRMPRIMEGLES::~CRendererDRMPRIMEGLES()
 {
   Flush(false);
+
+  auto winSystem = CServiceBroker::GetWinSystem();
+
+  if (!winSystem)
+    return;
+
+  winSystem->SetHDR(nullptr);
 }
 
 CBaseRenderer* CRendererDRMPRIMEGLES::Create(CVideoBuffer* buffer)
@@ -125,6 +132,15 @@ bool CRendererDRMPRIMEGLES::Configure(const VideoPicture& picture,
 
   m_clearColour = winSystem->UseLimitedColor() ? (16.0f / 0xff) : 0.0f;
 
+  if (picture.hasDisplayMetadata || picture.hasLightMetadata)
+  {
+    // note that this doesn't mean much right now as we blindly set HDR metadata if the
+    // connector property supports it. Can we improve this with a callback after a modeset?
+    m_passthroughHDR = winSystem->SetHDR(&picture);
+    CLog::Log(LOGDEBUG, "CRendererDRMPRIMEGLES::Configure - HDR passthrough: {}",
+              m_passthroughHDR ? "on" : "off");
+  }
+
   m_configured = true;
   return true;
 }
@@ -298,7 +314,7 @@ void CRendererDRMPRIMEGLES::RenderUpdate(
 
   bool toneMap = false;
 
-  if (m_videoSettings.m_ToneMapMethod != VS_TONEMAPMETHOD_OFF)
+  if (!m_passthroughHDR && m_videoSettings.m_ToneMapMethod != VS_TONEMAPMETHOD_OFF)
   {
     if (buf.hasLightMetadata || (buf.hasDisplayMetadata && buf.displayMetadata.has_luminance))
     {
@@ -354,7 +370,8 @@ void CRendererDRMPRIMEGLES::RenderUpdate(
   if (m_reloadShaders)
   {
     m_progressiveShader = std::make_unique<Shaders::GLES::YUV2RGBProgressiveShader>(
-        m_shaderFormat, AVColorPrimaries::AVCOL_PRI_BT709, m_srcPrimaries, m_toneMap, VS_TONEMAPMETHOD_OFF);
+        m_shaderFormat, m_passthroughHDR ? m_srcPrimaries : AVColorPrimaries::AVCOL_PRI_BT709,
+        m_srcPrimaries, m_toneMap, m_videoSettings.m_ToneMapMethod);
 
     m_progressiveShader->CompileAndLink();
 
diff --git a/xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererDRMPRIMEGLES.h b/xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererDRMPRIMEGLES.h
index 1308e3f284e75..e3223cba9a2c2 100644
--- a/xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererDRMPRIMEGLES.h
+++ b/xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererDRMPRIMEGLES.h
@@ -86,6 +86,7 @@ class CRendererDRMPRIMEGLES : public CBaseRenderer
   bool m_fullRange;
   AVColorPrimaries m_srcPrimaries;
   bool m_toneMap = false;
+  bool m_passthroughHDR = false;
 
   struct BUFFER
   {
diff --git a/xbmc/windowing/gbm/WinSystemGbm.h b/xbmc/windowing/gbm/WinSystemGbm.h
index ab156be786b07..5c9dab4d7031b 100644
--- a/xbmc/windowing/gbm/WinSystemGbm.h
+++ b/xbmc/windowing/gbm/WinSystemGbm.h
@@ -18,8 +18,38 @@
 
 #include <utility>
 
+#include <drm_mode.h>
 #include <gbm.h>
 
+#ifndef HAVE_HDR_OUTPUT_METADATA
+// HDR structs is copied from linux include/linux/hdmi.h
+struct hdr_metadata_infoframe
+{
+  uint8_t eotf;
+  uint8_t metadata_type;
+  struct
+  {
+    uint16_t x, y;
+  } display_primaries[3];
+  struct
+  {
+    uint16_t x, y;
+  } white_point;
+  uint16_t max_display_mastering_luminance;
+  uint16_t min_display_mastering_luminance;
+  uint16_t max_cll;
+  uint16_t max_fall;
+};
+struct hdr_output_metadata
+{
+  uint32_t metadata_type;
+  union
+  {
+    struct hdr_metadata_infoframe hdmi_metadata_type1;
+  };
+};
+#endif
+
 class IDispResource;
 
 namespace KODI
@@ -124,6 +124,7 @@
 
 private:
   uint32_t m_hdr_blob_id = 0;
+  struct hdr_output_metadata m_hdr_metadata = {};
 };
 
 }
From 5e94f4c70119d2e21c4c49ce824b536f02d1a106 Mon Sep 17 00:00:00 2001
From: Lukas Rusak <lorusak@gmail.com>
Date: Mon, 23 Nov 2020 16:07:17 -0800
Subject: [PATCH] CRenderSystemGLES: query default framebuffer attachment sizes

---
 xbmc/rendering/gles/RenderSystemGLES.cpp | 19 +++++++++++++++++++
 1 file changed, 19 insertions(+)

diff --git a/xbmc/rendering/gles/RenderSystemGLES.cpp b/xbmc/rendering/gles/RenderSystemGLES.cpp
index d8557ff8a0f80..69540fed20379 100644
--- a/xbmc/rendering/gles/RenderSystemGLES.cpp
+++ b/xbmc/rendering/gles/RenderSystemGLES.cpp
@@ -105,6 +105,25 @@ bool CRenderSystemGLES::InitRenderSystem()
 
   CGUITextureGLES::Register();
 
+  GLint red;
+  glGetFramebufferAttachmentParameteriv(GL_FRAMEBUFFER, GL_BACK, GL_FRAMEBUFFER_ATTACHMENT_RED_SIZE,
+                                        &red);
+
+  GLint green;
+  glGetFramebufferAttachmentParameteriv(GL_FRAMEBUFFER, GL_BACK,
+                                        GL_FRAMEBUFFER_ATTACHMENT_GREEN_SIZE, &green);
+
+  GLint blue;
+  glGetFramebufferAttachmentParameteriv(GL_FRAMEBUFFER, GL_BACK,
+                                        GL_FRAMEBUFFER_ATTACHMENT_BLUE_SIZE, &blue);
+
+  GLint alpha;
+  glGetFramebufferAttachmentParameteriv(GL_FRAMEBUFFER, GL_BACK,
+                                        GL_FRAMEBUFFER_ATTACHMENT_ALPHA_SIZE, &alpha);
+
+  CLog::Log(LOGDEBUG, "OpenGL(ES): framebuffer size: red: {} green: {} blue: {} alpha: {}", red,
+            green, blue, alpha);
+
   return true;
 }
 
diff --git a/xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodecDRMPRIME.cpp b/xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodecDRMPRIME.cpp
index ec2016cc060ff..bed1bfa88ed64 100644
--- a/xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodecDRMPRIME.cpp
+++ b/xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodecDRMPRIME.cpp
@@ -308,6 +308,7 @@ bool CDVDVideoCodecDRMPRIME::Open(CDVDStreamInfo& hints, CDVDCodecOptions& optio
   m_pCodecContext->opaque = static_cast<void*>(this);
   m_pCodecContext->get_format = GetFormat;
   m_pCodecContext->get_buffer2 = GetBuffer;
+  m_pCodecContext->thread_safe_callbacks = 1;
   m_pCodecContext->codec_tag = hints.codec_tag;
   m_pCodecContext->coded_width = hints.width;
   m_pCodecContext->coded_height = hints.height;
diff --git a/xbmc/interfaces/python/XBPython.cpp b/xbmc/interfaces/python/XBPython.cpp
index ee8ed93ff56ed..21f2de1382735 100644
--- a/xbmc/interfaces/python/XBPython.cpp
+++ b/xbmc/interfaces/python/XBPython.cpp
@@ -470,7 +470,7 @@ bool XBPython::OnScriptInitialized(ILanguageInvoker* invoker)
   {
     // Darwin packs .pyo files, we need PYTHONOPTIMIZE on in order to load them.
     // linux built with unified builds only packages the pyo files so need it
-#if defined(TARGET_DARWIN) || defined(TARGET_LINUX)
+#if defined(TARGET_DARWIN)
     setenv("PYTHONOPTIMIZE", "1", 1);
 #endif
     // Info about interesting python envvars available   
