diff -uNr a/xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodecDRMPRIME.cpp b/xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodecDRMPRIME.cpp
--- a/xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodecDRMPRIME.cpp	2024-03-02 22:30:13.000000000 +0100
+++ b/xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodecDRMPRIME.cpp	2024-03-06 17:18:14.047199153 +0100
@@ -476,7 +476,13 @@
   if (aspect_ratio <= 0.0)
     aspect_ratio =
         static_cast<double>(pVideoPicture->iWidth) / static_cast<double>(pVideoPicture->iHeight);
-
+  
+  if (m_pCodecContext->codec_id == AV_CODEC_ID_H264 &&
+      		       aspect_ratio != m_hints.aspect)
+  {
+  	aspect_ratio = m_hints.aspect;
+  }
+  
   if (m_DAR != aspect_ratio)
   {
     m_DAR = aspect_ratio;
