From 9273d123ecc0b128ff88bd3707a3d9fa3f3437b5 Mon Sep 17 00:00:00 2001
From: Preclikos <preclikos@preclikos.cz>
Date: Wed, 27 Oct 2021 23:43:10 +0200
Subject: [PATCH] parsers: increase DVBSUB max timediff (DTS/PCR)

---
 src/parsers/parsers.c | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/src/parsers/parsers.c b/src/parsers/parsers.c
index 7c20d01d21..e13cb5dbf2 100644
--- a/src/parsers/parsers.c
+++ b/src/parsers/parsers.c
@@ -112,7 +112,7 @@ parser_rstlog(parser_t *t, th_pkt_t *pkt)
 static void
 parser_deliver(parser_t *t, parser_es_t *st, th_pkt_t *pkt)
 {
-  int64_t d, diff;
+  int64_t d, diff, dts_increase;
 
   assert(pkt->pkt_type == st->es_type);
 
@@ -126,8 +126,9 @@ parser_deliver(parser_t *t, parser_es_t *st, th_pkt_t *pkt)
   }
 
   diff = st->es_type == SCT_DVBSUB ? 8*90000 : 6*90000;
-  d = pts_diff(pkt->pkt_pcr, (pkt->pkt_dts + 30000) & PTS_MASK);
-
+  dts_increase = st->es_type == SCT_DVBSUB ? 120000 : 30000;
+  d = pts_diff(pkt->pkt_pcr, (pkt->pkt_dts + dts_increase) & PTS_MASK);
+  
   if (d > diff || d == PTS_UNSET) {
     if (d != PTS_UNSET && tvhlog_limit(&st->es_pcr_log, 2))
       tvhwarn(LS_PARSER, "%s: DTS and PCR diff is very big (%"PRId64")",

