diff -uNr a/Makefile b/Makefile
--- a/Makefile	2023-03-13 10:05:24.360809000 +0100
+++ b/Makefile	2023-03-13 10:26:47.360431000 +0100
@@ -410,13 +410,28 @@
       else
          GLES3 = 1
       endif
+   else ifneq (,$(findstring RK3588,$(platform)))
+      CPUFLAGS += -march=armv8.2-a+crc+crypto -mtune=cortex-a76.cortex-a55
+      GLES3 = 1
+    
+  
+        
+         
    endif
 
-   GL_LIB := -lGLESv2
-   HAVE_NEON = 1
-   WITH_DYNAREC=arm
-   COREFLAGS += -DUSE_GENERIC_GLESV2 -DOS_LINUX
-   ASFLAGS = -f elf -d ELF_TYPE
+   ifeq ($(ARCH), aarch64)
+      GL_LIB := -lGLESv2
+      
+      WITH_DYNAREC=aarch64
+      COREFLAGS +=  -DOS_LINUX
+      ASFLAGS = -f elf64 -d ELF_TYPE
+   else
+      GL_LIB := -lGLESv2
+      HAVE_NEON = 1
+      WITH_DYNAREC=arm
+      COREFLAGS += -DUSE_GENERIC_GLESV2 -DOS_LINUX
+      ASFLAGS = -f elf -d ELF_TYPE
+   endif
 
 # OS X
 else ifneq (,$(findstring osx,$(platform)))
