diff -uNr a/xbmc/peripherals/devices/PeripheralCecAdapter.cpp b/xbmc/peripherals/devices/PeripheralCecAdapter.cpp
--- a/xbmc/peripherals/devices/PeripheralCecAdapter.cpp	2023-01-30 09:23:58.468330000 +0100
+++ b/xbmc/peripherals/devices/PeripheralCecAdapter.cpp	2023-01-31 13:11:54.314273654 +0100
@@ -176,6 +176,16 @@
         StandbyDevices();
     }
   }
+  else if (flag == ANNOUNCEMENT::GUI && sender == CAnnouncementManager::ANNOUNCEMENT_SENDER &&
+           message == "OnDPMSActivated")
+  {
+    std::unique_lock<CCriticalSection> lock(m_critSection);
+    const auto& components = CServiceBroker::GetAppComponents();
+    if (m_cecAdapter->IsLibCECActiveSource())
+    {
+        system("cec-ctl --standby -t0 > /dev/null");
+    }
+  }  
   else if (flag == ANNOUNCEMENT::System && sender == CAnnouncementManager::ANNOUNCEMENT_SENDER &&
            message == "OnSleep")
   {
