From d27b6ba1f5cd698df920a16d394e13000bd0f6fa Mon Sep 17 00:00:00 2001
From: joern-p <ahap1@web.de>
Date: Tue, 19 May 2020 13:26:32 +0200
Subject: [PATCH] ttz

---
 xbmc/Application.cpp | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/xbmc/Application.cpp b/xbmc/Application.cpp
index fde7fdaee31a..c283c49ebf44 100644
--- a/xbmc/Application.cpp
+++ b/xbmc/Application.cpp
@@ -4111,9 +4111,12 @@ void CApplication::ProcessSlow()
   {
     CJobManager::GetInstance().UnPauseJobs();
   }
-
+  
+  if (!m_appPlayer.IsPausedPlayback())
+  {
   // Check if we need to activate the screensaver / DPMS.
   CheckScreenSaverAndDPMS();
+  }
 
   // Check if we need to shutdown (if enabled).
 #if defined(TARGET_DARWIN)

