diff -uNr a/drivers/dma-buf/dma-buf.c b/drivers/dma-buf/dma-buf.c
--- a/drivers/dma-buf/dma-buf.c	2025-02-05 13:54:15.000000000 +0100
+++ b/drivers/dma-buf/dma-buf.c	2025-02-07 14:41:34.632744833 +0100
@@ -1532,30 +1532,6 @@
 }
 EXPORT_SYMBOL_NS_GPL(dma_buf_begin_cpu_access, DMA_BUF);
 
-int dma_buf_begin_cpu_access_partial(struct dma_buf *dmabuf,
-				     enum dma_data_direction direction,
-				     unsigned int offset, unsigned int len)
-{
-	int ret = 0;
-
-	if (WARN_ON(!dmabuf))
-		return -EINVAL;
-
-	if (dmabuf->ops->begin_cpu_access_partial)
-		ret = dmabuf->ops->begin_cpu_access_partial(dmabuf, direction,
-							    offset, len);
-
-	/* Ensure that all fences are waited upon - but we first allow
-	 * the native handler the chance to do so more efficiently if it
-	 * chooses. A double invocation here will be reasonably cheap no-op.
-	 */
-	if (ret == 0)
-		ret = __dma_buf_begin_cpu_access(dmabuf, direction);
-
-	return ret;
-}
-EXPORT_SYMBOL_GPL(dma_buf_begin_cpu_access_partial);
-
 /**
  * dma_buf_end_cpu_access - Must be called after accessing a dma_buf from the
  * cpu in the kernel context. Calls end_cpu_access to allow exporter-specific
@@ -1584,21 +1560,6 @@
 }
 EXPORT_SYMBOL_NS_GPL(dma_buf_end_cpu_access, DMA_BUF);
 
-int dma_buf_end_cpu_access_partial(struct dma_buf *dmabuf,
-				   enum dma_data_direction direction,
-				   unsigned int offset, unsigned int len)
-{
-	int ret = 0;
-
-	WARN_ON(!dmabuf);
-
-	if (dmabuf->ops->end_cpu_access_partial)
-		ret = dmabuf->ops->end_cpu_access_partial(dmabuf, direction,
-							  offset, len);
-
-	return ret;
-}
-EXPORT_SYMBOL_GPL(dma_buf_end_cpu_access_partial);
 
 /**
  * dma_buf_mmap - Setup up a userspace mmap with the given vma
diff -uNr a/drivers/dma-buf/dma-heap.c b/drivers/dma-buf/dma-heap.c
--- a/drivers/dma-buf/dma-heap.c	2025-02-05 13:54:15.000000000 +0100
+++ b/drivers/dma-buf/dma-heap.c	2025-02-07 14:42:25.432111172 +0100
@@ -41,7 +41,6 @@
 	dev_t heap_devt;
 	struct list_head list;
 	struct cdev heap_cdev;
-	struct device *heap_dev;
 };
 
 static LIST_HEAD(heap_list);
@@ -50,44 +49,7 @@
 static struct class *dma_heap_class;
 static DEFINE_XARRAY_ALLOC(dma_heap_minors);
 
-struct dma_heap *dma_heap_find(const char *name)
-{
-	struct dma_heap *h;
-
-	mutex_lock(&heap_list_lock);
-	list_for_each_entry(h, &heap_list, list) {
-		if (!strcmp(h->name, name)) {
-			mutex_unlock(&heap_list_lock);
-			return h;
-		}
-	}
-	mutex_unlock(&heap_list_lock);
-	return NULL;
-}
-EXPORT_SYMBOL_GPL(dma_heap_find);
-
-struct dma_buf *dma_heap_buffer_alloc(struct dma_heap *heap, size_t len,
-				      unsigned int fd_flags,
-				      unsigned int heap_flags)
-{
-	if (fd_flags & ~DMA_HEAP_VALID_FD_FLAGS)
-		return ERR_PTR(-EINVAL);
-
-	if (heap_flags & ~DMA_HEAP_VALID_HEAP_FLAGS)
-		return ERR_PTR(-EINVAL);
-	/*
-	 * Allocations from all heaps have to begin
-	 * and end on page boundaries.
-	 */
-	len = PAGE_ALIGN(len);
-	if (!len)
-		return ERR_PTR(-EINVAL);
-
-	return heap->ops->allocate(heap, len, fd_flags, heap_flags);
-}
-EXPORT_SYMBOL_GPL(dma_heap_buffer_alloc);
-
-static int dma_heap_bufferfd_alloc(struct dma_heap *heap, size_t len,
+static int dma_heap_buffer_alloc(struct dma_heap *heap, size_t len,
 				 unsigned int fd_flags,
 				 unsigned int heap_flags)
 {
@@ -146,7 +108,7 @@
 	if (heap_allocation->heap_flags & ~DMA_HEAP_VALID_HEAP_FLAGS)
 		return -EINVAL;
 
-	fd = dma_heap_bufferfd_alloc(heap, heap_allocation->len,
+	fd = dma_heap_buffer_alloc(heap, heap_allocation->len,
 				   heap_allocation->fd_flags,
 				   heap_allocation->heap_flags);
 	if (fd < 0)
@@ -260,19 +222,6 @@
 }
 
 /**
- * dma_heap_get_dev() - get device struct for the heap
- * @heap: DMA-Heap to retrieve device struct from
- *
- * Returns:
- * The device struct for the heap.
- */
-struct device *dma_heap_get_dev(struct dma_heap *heap)
-{
-	return heap->heap_dev;
-}
-EXPORT_SYMBOL_GPL(dma_heap_get_dev);
-
-/**
  * dma_heap_get_name() - get heap name
  * @heap: DMA-Heap to retrieve private data for
  *
@@ -340,9 +289,6 @@
 		goto err2;
 	}
 
-	heap->heap_dev = dev_ret;
-	heap->heap_dev = get_device(heap->heap_dev);
-
 	mutex_lock(&heap_list_lock);
 	/* check the name is unique */
 	list_for_each_entry(h, &heap_list, list) {
@@ -351,7 +297,6 @@
 			pr_err("dma_heap: Already registered heap named %s\n",
 			       exp_info->name);
 			err_ret = ERR_PTR(-EINVAL);
-			put_device(heap->heap_dev);
 			goto err3;
 		}
 	}
diff -uNr a/drivers/dma-buf/heaps/system_heap.c b/drivers/dma-buf/heaps/system_heap.c
--- a/drivers/dma-buf/heaps/system_heap.c	2025-02-05 13:54:15.000000000 +0100
+++ b/drivers/dma-buf/heaps/system_heap.c	2025-01-05 10:09:31.000000000 +0100
@@ -22,7 +22,6 @@
 #include <linux/vmalloc.h>
 
 static struct dma_heap *sys_heap;
-static struct dma_heap *sys_uncached_heap;
 
 struct system_heap_buffer {
 	struct dma_heap *heap;
@@ -32,8 +31,6 @@
 	struct sg_table sg_table;
 	int vmap_cnt;
 	void *vaddr;
-
-	bool uncached;
 };
 
 struct dma_heap_attachment {
@@ -41,8 +38,6 @@
 	struct sg_table *table;
 	struct list_head list;
 	bool mapped;
-
-	bool uncached;
 };
 
 #define LOW_ORDER_GFP (GFP_HIGHUSER | __GFP_ZERO)
@@ -105,7 +100,7 @@
 	a->dev = attachment->dev;
 	INIT_LIST_HEAD(&a->list);
 	a->mapped = false;
-	a->uncached = buffer->uncached;
+
 	attachment->priv = a;
 
 	mutex_lock(&buffer->lock);
@@ -135,13 +130,9 @@
 {
 	struct dma_heap_attachment *a = attachment->priv;
 	struct sg_table *table = a->table;
-	int attr = attachment->dma_map_attrs;
 	int ret;
 
-	if (a->uncached)
-		attr |= DMA_ATTR_SKIP_CPU_SYNC;
-
-	ret = dma_map_sgtable(attachment->dev, table, direction, attr);
+	ret = dma_map_sgtable(attachment->dev, table, direction, 0);
 	if (ret)
 		return ERR_PTR(ret);
 
@@ -154,12 +145,9 @@
 				      enum dma_data_direction direction)
 {
 	struct dma_heap_attachment *a = attachment->priv;
-	int attr = attachment->dma_map_attrs;
 
-	if (a->uncached)
-		attr |= DMA_ATTR_SKIP_CPU_SYNC;
 	a->mapped = false;
-	dma_unmap_sgtable(attachment->dev, table, direction, attr);
+	dma_unmap_sgtable(attachment->dev, table, direction, 0);
 }
 
 static int system_heap_dma_buf_begin_cpu_access(struct dma_buf *dmabuf,
@@ -173,12 +161,10 @@
 	if (buffer->vmap_cnt)
 		invalidate_kernel_vmap_range(buffer->vaddr, buffer->len);
 
-	if (!buffer->uncached) {
-		list_for_each_entry(a, &buffer->attachments, list) {
-			if (!a->mapped)
-				continue;
-			dma_sync_sgtable_for_cpu(a->dev, a->table, direction);
-		}
+	list_for_each_entry(a, &buffer->attachments, list) {
+		if (!a->mapped)
+			continue;
+		dma_sync_sgtable_for_cpu(a->dev, a->table, direction);
 	}
 	mutex_unlock(&buffer->lock);
 
@@ -196,115 +182,14 @@
 	if (buffer->vmap_cnt)
 		flush_kernel_vmap_range(buffer->vaddr, buffer->len);
 
-	if (!buffer->uncached) {
-		list_for_each_entry(a, &buffer->attachments, list) {
-			if (!a->mapped)
-				continue;
-			dma_sync_sgtable_for_device(a->dev, a->table, direction);
-		}
-	}
-	mutex_unlock(&buffer->lock);
-
-	return 0;
-}
-
-static int system_heap_sgl_sync_range(struct device *dev,
-				      struct sg_table *sgt,
-				      unsigned int offset,
-				      unsigned int length,
-				      enum dma_data_direction dir,
-				      bool for_cpu)
-{
-	struct scatterlist *sg;
-	unsigned int len = 0;
-	dma_addr_t sg_dma_addr;
-	int i;
-
-	for_each_sgtable_sg(sgt, sg, i) {
-		unsigned int sg_offset, sg_left, size = 0;
-
-		sg_dma_addr = sg_phys(sg);
-
-		len += sg->length;
-		if (len <= offset)
+	list_for_each_entry(a, &buffer->attachments, list) {
+		if (!a->mapped)
 			continue;
-
-		sg_left = len - offset;
-		sg_offset = sg->length - sg_left;
-
-		size = (length < sg_left) ? length : sg_left;
-		if (for_cpu)
-			dma_sync_single_range_for_cpu(dev, sg_dma_addr,
-						      sg_offset, size, dir);
-		else
-			dma_sync_single_range_for_device(dev, sg_dma_addr,
-							 sg_offset, size, dir);
-
-		offset += size;
-		length -= size;
-
-		if (length == 0)
-			break;
+		dma_sync_sgtable_for_device(a->dev, a->table, direction);
 	}
-
-	return 0;
-}
-
-static int __maybe_unused
-system_heap_dma_buf_begin_cpu_access_partial(struct dma_buf *dmabuf,
-					     enum dma_data_direction direction,
-					     unsigned int offset,
-					     unsigned int len)
-{
-	struct system_heap_buffer *buffer = dmabuf->priv;
-	struct dma_heap *heap = buffer->heap;
-	struct sg_table *table = &buffer->sg_table;
-	int ret;
-
-	if (direction == DMA_TO_DEVICE)
-		return 0;
-
-	mutex_lock(&buffer->lock);
-	if (buffer->vmap_cnt)
-		invalidate_kernel_vmap_range(buffer->vaddr, buffer->len);
-
-	if (buffer->uncached) {
-		mutex_unlock(&buffer->lock);
-		return 0;
-	}
-
-	ret = system_heap_sgl_sync_range(dma_heap_get_dev(heap), table,
-					 offset, len, direction, true);
 	mutex_unlock(&buffer->lock);
 
-	return ret;
-}
-
-static int __maybe_unused
-system_heap_dma_buf_end_cpu_access_partial(struct dma_buf *dmabuf,
-					   enum dma_data_direction direction,
-					   unsigned int offset,
-					   unsigned int len)
-{
-	struct system_heap_buffer *buffer = dmabuf->priv;
-	struct dma_heap *heap = buffer->heap;
-	struct sg_table *table = &buffer->sg_table;
-	int ret;
-
-	mutex_lock(&buffer->lock);
-	if (buffer->vmap_cnt)
-		flush_kernel_vmap_range(buffer->vaddr, buffer->len);
-
-	if (buffer->uncached) {
-		mutex_unlock(&buffer->lock);
-		return 0;
-	}
-
-	ret = system_heap_sgl_sync_range(dma_heap_get_dev(heap), table,
-					 offset, len, direction, false);
-	mutex_unlock(&buffer->lock);
-
-	return ret;
+	return 0;
 }
 
 static int system_heap_mmap(struct dma_buf *dmabuf, struct vm_area_struct *vma)
@@ -315,9 +200,6 @@
 	struct sg_page_iter piter;
 	int ret;
 
-	if (buffer->uncached)
-		vma->vm_page_prot = pgprot_writecombine(vma->vm_page_prot);
-
 	for_each_sgtable_page(table, &piter, vma->vm_pgoff) {
 		struct page *page = sg_page_iter_page(&piter);
 
@@ -339,21 +221,17 @@
 	struct page **pages = vmalloc(sizeof(struct page *) * npages);
 	struct page **tmp = pages;
 	struct sg_page_iter piter;
-	pgprot_t pgprot = PAGE_KERNEL;
 	void *vaddr;
 
 	if (!pages)
 		return ERR_PTR(-ENOMEM);
 
-	if (buffer->uncached)
-		pgprot = pgprot_writecombine(PAGE_KERNEL);
-
 	for_each_sgtable_page(table, &piter, 0) {
 		WARN_ON(tmp - pages >= npages);
 		*tmp++ = sg_page_iter_page(&piter);
 	}
 
-	vaddr = vmap(pages, npages, VM_MAP, pgprot);
+	vaddr = vmap(pages, npages, VM_MAP, PAGE_KERNEL);
 	vfree(pages);
 
 	if (!vaddr)
@@ -427,10 +305,6 @@
 	.unmap_dma_buf = system_heap_unmap_dma_buf,
 	.begin_cpu_access = system_heap_dma_buf_begin_cpu_access,
 	.end_cpu_access = system_heap_dma_buf_end_cpu_access,
-#ifdef CONFIG_DMABUF_PARTIAL
-	.begin_cpu_access_partial = system_heap_dma_buf_begin_cpu_access_partial,
-	.end_cpu_access_partial = system_heap_dma_buf_end_cpu_access_partial,
-#endif
 	.mmap = system_heap_mmap,
 	.vmap = system_heap_vmap,
 	.vunmap = system_heap_vunmap,
@@ -457,11 +331,10 @@
 	return NULL;
 }
 
-static struct dma_buf *system_heap_do_allocate(struct dma_heap *heap,
-					       unsigned long len,
-					       unsigned long fd_flags,
-					       unsigned long heap_flags,
-					       bool uncached)
+static struct dma_buf *system_heap_allocate(struct dma_heap *heap,
+					    unsigned long len,
+					    unsigned long fd_flags,
+					    unsigned long heap_flags)
 {
 	struct system_heap_buffer *buffer;
 	DEFINE_DMA_BUF_EXPORT_INFO(exp_info);
@@ -482,7 +355,6 @@
 	mutex_init(&buffer->lock);
 	buffer->heap = heap;
 	buffer->len = len;
-	buffer->uncached = uncached;
 
 	INIT_LIST_HEAD(&pages);
 	i = 0;
@@ -528,18 +400,6 @@
 		ret = PTR_ERR(dmabuf);
 		goto free_pages;
 	}
-
-	/*
-	 * For uncached buffers, we need to initially flush cpu cache, since
-	 * the __GFP_ZERO on the allocation means the zeroing was done by the
-	 * cpu and thus it is likely cached. Map (and implicitly flush) and
-	 * unmap it now so we don't get corruption later on.
-	 */
-	if (buffer->uncached) {
-		dma_map_sgtable(dma_heap_get_dev(heap), table, DMA_BIDIRECTIONAL, 0);
-		dma_unmap_sgtable(dma_heap_get_dev(heap), table, DMA_BIDIRECTIONAL, 0);
-	}
-
 	return dmabuf;
 
 free_pages:
@@ -557,39 +417,10 @@
 	return ERR_PTR(ret);
 }
 
-static struct dma_buf *system_heap_allocate(struct dma_heap *heap,
-					    unsigned long len,
-					    unsigned long fd_flags,
-					    unsigned long heap_flags)
-{
-	return system_heap_do_allocate(heap, len, fd_flags, heap_flags, false);
-}
 static const struct dma_heap_ops system_heap_ops = {
 	.allocate = system_heap_allocate,
 };
 
-static struct dma_buf *system_uncached_heap_allocate(struct dma_heap *heap,
-						     unsigned long len,
-						     unsigned long fd_flags,
-						     unsigned long heap_flags)
-{
-	return system_heap_do_allocate(heap, len, fd_flags, heap_flags, true);
-}
-
-/* Dummy function to be used until we can call coerce_mask_and_coherent */
-static struct dma_buf *system_uncached_heap_not_initialized(struct dma_heap *heap,
-							    unsigned long len,
-							    unsigned long fd_flags,
-							    unsigned long heap_flags)
-{
-	return ERR_PTR(-EBUSY);
-}
-
-static struct dma_heap_ops system_uncached_heap_ops = {
-	/* After system_heap_create is complete, we will swap this */
-	.allocate = system_uncached_heap_not_initialized,
-};
-
 static int system_heap_create(void)
 {
 	struct dma_heap_export_info exp_info;
@@ -602,18 +433,6 @@
 	if (IS_ERR(sys_heap))
 		return PTR_ERR(sys_heap);
 
-	exp_info.name = "system-uncached";
-	exp_info.ops = &system_uncached_heap_ops;
-	exp_info.priv = NULL;
-
-	sys_uncached_heap = dma_heap_add(&exp_info);
-	if (IS_ERR(sys_uncached_heap))
-		return PTR_ERR(sys_uncached_heap);
-
-	dma_coerce_mask_and_coherent(dma_heap_get_dev(sys_uncached_heap), DMA_BIT_MASK(64));
-	mb(); /* make sure we only set allocate after dma_mask is set */
-	system_uncached_heap_ops.allocate = system_uncached_heap_allocate;
-
 	return 0;
 }
 module_init(system_heap_create);
diff -uNr a/include/linux/dma-buf.h b/include/linux/dma-buf.h
--- a/include/linux/dma-buf.h	2025-02-05 13:54:15.000000000 +0100
+++ b/include/linux/dma-buf.h	2025-01-05 10:09:31.000000000 +0100
@@ -230,41 +230,6 @@
 	int (*begin_cpu_access)(struct dma_buf *, enum dma_data_direction);
 
 	/**
-	 * @begin_cpu_access_partial:
-	 *
-	 * This is called from dma_buf_begin_cpu_access_partial() and allows the
-	 * exporter to ensure that the memory specified in the range is
-	 * available for cpu access - the exporter might need to allocate or
-	 * swap-in and pin the backing storage.
-	 * The exporter also needs to ensure that cpu access is
-	 * coherent for the access direction. The direction can be used by the
-	 * exporter to optimize the cache flushing, i.e. access with a different
-	 * direction (read instead of write) might return stale or even bogus
-	 * data (e.g. when the exporter needs to copy the data to temporary
-	 * storage).
-	 *
-	 * This callback is optional.
-	 *
-	 * FIXME: This is both called through the DMA_BUF_IOCTL_SYNC command
-	 * from userspace (where storage shouldn't be pinned to avoid handing
-	 * de-factor mlock rights to userspace) and for the kernel-internal
-	 * users of the various kmap interfaces, where the backing storage must
-	 * be pinned to guarantee that the atomic kmap calls can succeed. Since
-	 * there's no in-kernel users of the kmap interfaces yet this isn't a
-	 * real problem.
-	 *
-	 * Returns:
-	 *
-	 * 0 on success or a negative error code on failure. This can for
-	 * example fail when the backing storage can't be allocated. Can also
-	 * return -ERESTARTSYS or -EINTR when the call has been interrupted and
-	 * needs to be restarted.
-	 */
-	int (*begin_cpu_access_partial)(struct dma_buf *dmabuf,
-					enum dma_data_direction,
-					unsigned int offset, unsigned int len);
-
-	/**
 	 * @end_cpu_access:
 	 *
 	 * This is called from dma_buf_end_cpu_access() when the importer is
@@ -282,28 +247,6 @@
 	int (*end_cpu_access)(struct dma_buf *, enum dma_data_direction);
 
 	/**
-	 * @end_cpu_access_partial:
-	 *
-	 * This is called from dma_buf_end_cpu_access_partial() when the
-	 * importer is done accessing the CPU. The exporter can use to limit
-	 * cache flushing to only the range specefied and to unpin any
-	 * resources pinned in @begin_cpu_access_umapped.
-	 * The result of any dma_buf kmap calls after end_cpu_access_partial is
-	 * undefined.
-	 *
-	 * This callback is optional.
-	 *
-	 * Returns:
-	 *
-	 * 0 on success or a negative error code on failure. Can return
-	 * -ERESTARTSYS or -EINTR when the call has been interrupted and needs
-	 * to be restarted.
-	 */
-	int (*end_cpu_access_partial)(struct dma_buf *dmabuf,
-				      enum dma_data_direction,
-				      unsigned int offset, unsigned int len);
-
-	/**
 	 * @mmap:
 	 *
 	 * This callback is used by the dma_buf_mmap() function
@@ -580,8 +523,6 @@
  * @importer_ops: importer operations for this attachment, if provided
  * dma_buf_map/unmap_attachment() must be called with the dma_resv lock held.
  * @importer_priv: importer specific attachment data.
- * @dma_map_attrs: DMA attributes to be used when the exporter maps the buffer
- * through dma_buf_map_attachment.
  *
  * This structure holds the attachment information between the dma_buf buffer
  * and its user device(s). The list contains one attachment struct per device
@@ -602,7 +543,6 @@
 	const struct dma_buf_attach_ops *importer_ops;
 	void *importer_priv;
 	void *priv;
-	unsigned long dma_map_attrs;
 };
 
 /**
@@ -706,14 +646,8 @@
 void dma_buf_move_notify(struct dma_buf *dma_buf);
 int dma_buf_begin_cpu_access(struct dma_buf *dma_buf,
 			     enum dma_data_direction dir);
-int dma_buf_begin_cpu_access_partial(struct dma_buf *dma_buf,
-				     enum dma_data_direction dir,
-				     unsigned int offset, unsigned int len);
 int dma_buf_end_cpu_access(struct dma_buf *dma_buf,
 			   enum dma_data_direction dir);
-int dma_buf_end_cpu_access_partial(struct dma_buf *dma_buf,
-				     enum dma_data_direction dir,
-				     unsigned int offset, unsigned int len);
 struct sg_table *
 dma_buf_map_attachment_unlocked(struct dma_buf_attachment *attach,
 				enum dma_data_direction direction);
diff -uNr a/include/linux/dma-heap.h b/include/linux/dma-heap.h
--- a/include/linux/dma-heap.h	2025-02-05 13:54:15.000000000 +0100
+++ b/include/linux/dma-heap.h	2025-01-05 10:09:31.000000000 +0100
@@ -55,15 +55,6 @@
 void *dma_heap_get_drvdata(struct dma_heap *heap);
 
 /**
- * dma_heap_get_dev() - get device struct for the heap
- * @heap: DMA-Heap to retrieve device struct from
- *
- * Returns:
- * The device struct for the heap.
- */
-struct device *dma_heap_get_dev(struct dma_heap *heap);
-
-/**
  * dma_heap_get_name() - get heap name
  * @heap: DMA-Heap to retrieve private data for
  *
@@ -78,23 +69,4 @@
  */
 struct dma_heap *dma_heap_add(const struct dma_heap_export_info *exp_info);
 
-/**
- * dma_heap_find - Returns the registered dma_heap with the specified name
- * @name: Name of the heap to find
- *
- */
-struct dma_heap *dma_heap_find(const char *name);
-
-/**
- * dma_heap_buffer_alloc - Allocate dma-buf from a dma_heap
- * @heap:	dma_heap to allocate from
- * @len:	size to allocate
- * @fd_flags:	flags to set on returned dma-buf fd
- * @heap_flags:	flags to pass to the dma heap
- *
- * This is for internal dma-buf allocations only.
- */
-struct dma_buf *dma_heap_buffer_alloc(struct dma_heap *heap, size_t len,
-				      unsigned int fd_flags,
-				      unsigned int heap_flags);
 #endif /* _DMA_HEAPS_H */
