diff -uNr a/drivers/clk/rockchip/clk-rk3588.c b/drivers/clk/rockchip/clk-rk3588.c
--- a/drivers/clk/rockchip/clk-rk3588.c	2024-07-14 00:45:30.000000000 +0200
+++ b/drivers/clk/rockchip/clk-rk3588.c	2024-10-24 12:33:18.468141137 +0200
@@ -75,7 +75,11 @@
 	RK3588_PLL_RATE(1320000000, 2, 220, 1, 0),
 	RK3588_PLL_RATE(1200000000, 2, 200, 1, 0),
 	RK3588_PLL_RATE(1188000000, 2, 198, 1, 0),
+	RK3588_PLL_RATE(1186814000, 2, 198, 1, 52581),
+	RK3588_PLL_RATE(1186812000, 2, 198, 1, 52559),
+	RK3588_PLL_RATE(1109000000, 3, 554, 2, 32767),
 	RK3588_PLL_RATE(1100000000, 3, 550, 2, 0),
+	RK3588_PLL_RATE(1051000000, 3, 525, 2, 32767),
 	RK3588_PLL_RATE(1008000000, 2, 336, 2, 0),
 	RK3588_PLL_RATE(1000000000, 3, 500, 2, 0),
 	RK3588_PLL_RATE(983040000, 4, 655, 2, 23592),
@@ -86,11 +90,15 @@
 	RK3588_PLL_RATE(786432000, 2, 262, 2, 9437),
 	RK3588_PLL_RATE(786000000, 1, 131, 2, 0),
 	RK3588_PLL_RATE(785560000, 3, 393, 2, 51119),
+	RK3588_PLL_RATE(773000000, 2, 258, 2, 43690),
 	RK3588_PLL_RATE(722534400, 8, 963, 2, 24850),
+	RK3588_PLL_RATE(697000000, 2, 232, 2, 21845),
+	RK3588_PLL_RATE(604800000, 1, 101, 2, 52428),
 	RK3588_PLL_RATE(600000000, 2, 200, 2, 0),
 	RK3588_PLL_RATE(594000000, 1, 99, 2, 0),
 	RK3588_PLL_RATE(408000000, 2, 272, 3, 0),
 	RK3588_PLL_RATE(312000000, 2, 208, 3, 0),
+	RK3588_PLL_RATE(266580000, 1, 178, 4, 47185),
 	RK3588_PLL_RATE(216000000, 2, 288, 4, 0),
 	RK3588_PLL_RATE(96000000, 2, 256, 5, 0),
 	{ /* sentinel */ },
diff -uNr a/drivers/gpu/drm/bridge/synopsys/dw-hdmi-qp.c b/drivers/gpu/drm/bridge/synopsys/dw-hdmi-qp.c
--- a/drivers/gpu/drm/bridge/synopsys/dw-hdmi-qp.c	2024-07-14 00:45:30.000000000 +0200
+++ b/drivers/gpu/drm/bridge/synopsys/dw-hdmi-qp.c	2024-10-24 13:07:23.455830480 +0200
@@ -4291,10 +4291,13 @@
 
 	mutex_unlock(&hdmi->mutex);
 
+	if (hdmi->phy.ops->setup_hpd)
+		hdmi->phy.ops->setup_hpd(hdmi, hdmi->phy.data);
+
 	result = hdmi->phy.ops->read_hpd(hdmi, hdmi->phy.data);
 	if (result == connector_status_connected) {
 		mutex_lock(&hdmi->connector.dev->mode_config.mutex);
-		dw_hdmi_connector_get_modes(&hdmi->connector);
+		drm_helper_probe_single_connector_modes(&hdmi->connector, 9000, 9000);
 		mutex_unlock(&hdmi->connector.dev->mode_config.mutex);
 	}
 }
diff -uNr a/drivers/gpu/drm/drm_modes.c b/drivers/gpu/drm/drm_modes.c
--- a/drivers/gpu/drm/drm_modes.c	2024-07-14 00:45:30.000000000 +0200
+++ b/drivers/gpu/drm/drm_modes.c	2024-10-24 12:52:38.295294814 +0200
@@ -2012,6 +2012,7 @@
 	strncpy(out->name, in->name, DRM_DISPLAY_MODE_LEN);
 	out->name[DRM_DISPLAY_MODE_LEN-1] = 0;
 }
+EXPORT_SYMBOL_GPL(drm_mode_convert_to_umode);
 
 /**
  * drm_mode_convert_umode - convert a modeinfo into a drm_display_mode
@@ -2088,6 +2089,7 @@
 
 	return 0;
 }
+EXPORT_SYMBOL_GPL(drm_mode_convert_umode);
 
 /**
  * drm_mode_is_420_only - if a given videomode can be only supported in YCBCR420
diff -uNr a/drivers/gpu/drm/rockchip/rockchip_drm_drv.c b/drivers/gpu/drm/rockchip/rockchip_drm_drv.c
--- a/drivers/gpu/drm/rockchip/rockchip_drm_drv.c	2024-07-14 00:45:30.000000000 +0200
+++ b/drivers/gpu/drm/rockchip/rockchip_drm_drv.c	2024-10-24 12:59:56.771099547 +0200
@@ -1991,6 +1991,8 @@
 	.lastclose		= rockchip_drm_lastclose,
 	.open			= rockchip_drm_open,
 	.dumb_create		= rockchip_gem_dumb_create,
+	.prime_handle_to_fd	= drm_gem_prime_handle_to_fd,
+	.prime_fd_to_handle	= drm_gem_prime_fd_to_handle,
 	.gem_prime_import	= rockchip_drm_gem_prime_import,
 	.gem_prime_import_sg_table	= rockchip_gem_prime_import_sg_table,
 	.gem_prime_mmap		= drm_gem_prime_mmap,
diff -uNr a/drivers/gpu/drm/rockchip/rockchip_drm_drv.h b/drivers/gpu/drm/rockchip/rockchip_drm_drv.h
--- a/drivers/gpu/drm/rockchip/rockchip_drm_drv.h	2024-07-14 00:45:30.000000000 +0200
+++ b/drivers/gpu/drm/rockchip/rockchip_drm_drv.h	2024-10-24 13:01:15.822405078 +0200
@@ -506,7 +506,7 @@
 	void (*crtc_output_post_enable)(struct drm_crtc *crtc, int intf);
 	void (*crtc_output_pre_disable)(struct drm_crtc *crtc, int intf);
 	int (*crtc_set_color_bar)(struct drm_crtc *crtc, enum rockchip_color_bar_mode mode);
-	int (*set_aclk)(struct drm_crtc *crtc, enum rockchip_drm_vop_aclk_mode aclk_mode);
+	int (*set_aclk)(struct drm_crtc *crtc, enum rockchip_drm_vop_aclk_mode aclk_mode, struct dmcfreq_vop_info *vop_bw_info);
 	int (*get_crc)(struct drm_crtc *crtc);
 };
 
diff -uNr a/drivers/gpu/drm/rockchip/rockchip_drm_fb.c b/drivers/gpu/drm/rockchip/rockchip_drm_fb.c
--- a/drivers/gpu/drm/rockchip/rockchip_drm_fb.c	2024-07-14 00:45:30.000000000 +0200
+++ b/drivers/gpu/drm/rockchip/rockchip_drm_fb.c	2024-10-24 13:01:01.057530011 +0200
@@ -202,14 +202,14 @@
 			if (vop_bw_info->plane_num_4k || crtc_num > 1 ||
 			    crtc->state->adjusted_mode.crtc_hdisplay > 2560 ||
 			    crtc->state->adjusted_mode.crtc_vdisplay > 2560) {
-				funcs->set_aclk(crtc, ROCKCHIP_VOP_ACLK_ADVANCED_MODE);
+				funcs->set_aclk(crtc, ROCKCHIP_VOP_ACLK_ADVANCED_MODE, vop_bw_info);
 				priv->aclk_adjust_frame_num = 2;
 			} else {
 				if (priv->aclk_adjust_frame_num >= 1) {
-					funcs->set_aclk(crtc, ROCKCHIP_VOP_ACLK_ADVANCED_MODE);
+					funcs->set_aclk(crtc, ROCKCHIP_VOP_ACLK_ADVANCED_MODE, vop_bw_info);
 					priv->aclk_adjust_frame_num--;
 				} else {
-					funcs->set_aclk(crtc, ROCKCHIP_VOP_ACLK_NORMAL_MODE);
+					funcs->set_aclk(crtc, ROCKCHIP_VOP_ACLK_NORMAL_MODE, vop_bw_info);
 				}
 			}
 		}
diff -uNr a/drivers/gpu/drm/rockchip/rockchip_drm_vop2.c b/drivers/gpu/drm/rockchip/rockchip_drm_vop2.c
--- a/drivers/gpu/drm/rockchip/rockchip_drm_vop2.c	2024-07-14 00:45:30.000000000 +0200
+++ b/drivers/gpu/drm/rockchip/rockchip_drm_vop2.c	2024-10-24 13:18:37.013555293 +0200
@@ -968,6 +968,8 @@
 	u32 aclk_mode_rate[ROCKCHIP_VOP_ACLK_MAX_MODE];
 #endif
 
+	/* aclk auto cs div */
+	u32 csu_div;
 	/* must put at the end of the struct */
 	struct vop2_win win[];
 };
@@ -1306,6 +1308,99 @@
 	return NULL;
 }
 
+static u32 rk3562_vop2_get_csu_div(struct drm_crtc *crtc)
+{
+	struct vop2_video_port *vp = to_vop2_video_port(crtc);
+	unsigned long aclk_rate = 0, dclk_rate = 0;
+	u32 csu_div;
+
+	aclk_rate = clk_get_rate(vp->vop2->aclk);
+	dclk_rate = clk_get_rate(vp->dclk);
+	if (!dclk_rate)
+		return 0;
+
+	/* aclk > 1/2 * dclk */
+	csu_div = (aclk_rate - 1) * 2 / dclk_rate;
+
+	return csu_div;
+}
+
+static u32 rk3576_vop2_get_csu_div(struct drm_crtc *crtc, struct dmcfreq_vop_info *vop_bw_info)
+{
+	struct vop2_video_port *vp = to_vop2_video_port(crtc);
+	struct drm_display_mode *adjusted_mode = &crtc->state->adjusted_mode;
+	u32 csu_div;
+
+	if (vp->vop2->aclk_mode == ROCKCHIP_VOP_ACLK_NORMAL_MODE) {
+		if (adjusted_mode->crtc_clock < 250000) {
+			if (vop_bw_info->plane_num == 1)
+				csu_div = 3;
+			else if (vop_bw_info->plane_num == 2)
+				csu_div = 2;
+			else
+				csu_div = 1;
+		} else {
+			csu_div = 1;
+		}
+	} else {
+		csu_div = 1;
+	}
+
+	return csu_div;
+}
+
+static int vop2_set_aclk_rate(struct drm_crtc *crtc,
+			      enum rockchip_drm_vop_aclk_mode aclk_mode,
+			      struct dmcfreq_vop_info *vop_bw_info)
+{
+	struct vop2_video_port *vp = to_vop2_video_port(crtc);
+	struct vop2 *vop2 = vp->vop2;
+	struct drm_crtc *first_active_crtc = NULL;
+	int i = 0, ret = 0;
+	u32 csu_div = 0;
+
+	/* all vp/crtc share one vop aclk, so only need to set once */
+	for (i = 0; i < vop2->data->nr_vps; i++) {
+		if (vop2->vps[i].rockchip_crtc.crtc.state &&
+		    vop2->vps[i].rockchip_crtc.crtc.state->active) {
+			first_active_crtc = &vop2->vps[i].rockchip_crtc.crtc;
+			break;
+		}
+	}
+	if (first_active_crtc != crtc)
+		return 0;
+
+	vop2->aclk_target_freq = vop2->aclk_mode_rate[aclk_mode];
+
+#ifdef CONFIG_PM_DEVFREQ
+	if (vop2->devfreq) {
+		mutex_lock(&vop2->devfreq->lock);
+		ret = update_devfreq(vop2->devfreq);
+		mutex_unlock(&vop2->devfreq->lock);
+		if (ret)
+			dev_err(vop2->dev, "failed to set rate %lu\n", vop2->aclk_target_freq);
+	}
+#endif
+	vop2->aclk_mode = aclk_mode;
+
+	if (vop2->csu_aclk && vop_bw_info) {
+		if (vop2->version == VOP_VERSION_RK3562)
+			csu_div = rk3562_vop2_get_csu_div(crtc);
+		else
+			csu_div = rk3576_vop2_get_csu_div(crtc, vop_bw_info);
+		if (csu_div != vop2->csu_div) {
+			rockchip_csu_set_div(vop2->csu_aclk, csu_div);
+			rockchip_drm_dbg(vop2->dev, VOP_DEBUG_CLK,
+					 "Set aclk auto cs div from %d to %d, aclk rate:%ld, aclk mode:%d\n",
+					 vop2->csu_div, csu_div, clk_get_rate(vop2->aclk),
+					 vop2->aclk_mode);
+		}
+		vop2->csu_div = csu_div;
+	}
+
+	return 0;
+}
+
 static int vop2_clk_reset(struct reset_control *rstc)
 {
 	int ret;
@@ -1879,6 +1974,12 @@
 		vop2_power_domain_put(pd->parent);
 }
 
+static void vop2_power_domain_put_sync(struct vop2_power_domain *pd)
+{
+	vop2_power_domain_put(pd);
+	vop2_wait_power_domain_off(pd);
+}
+
 /*
  * Called if the pd ref_count reach 0 after 2.5
  * seconds.
@@ -2526,14 +2627,22 @@
 
 	if (is_vop3(vop2) && vop2->version != VOP_VERSION_RK3528) {
 		uint32_t vir_height = fb->height;
-		u8 block_w;
+		u8 block_w, block_h;
 
-		if (IS_ROCKCHIP_RFBC_MOD(fb->modifier))
+		if (IS_ROCKCHIP_RFBC_MOD(fb->modifier)) {
 			block_w = 64;
-		else if (fb->modifier & AFBC_FORMAT_MOD_BLOCK_SIZE_32x8)
+			block_h = 4;
+		} else if (fb->modifier & AFBC_FORMAT_MOD_BLOCK_SIZE_32x8) {
 			block_w = 32;
-		else
+			block_h = 8;
+		} else {
 			block_w = 16;
+			block_h = 16;
+		}
+		if (!IS_ALIGNED(vir_height, block_h)) {
+			DRM_WARN("FBC fb vir height[%d] should aligned as block height[%d]", vir_height, block_h);
+			vir_height = ALIGN(vir_height, block_h);
+		}
 
 		if (vpstate->xmirror_en) {
 			transform_tmp = ALIGN(act_xoffset + width, block_w);
@@ -2542,11 +2651,14 @@
 			transform_xoffset = act_xoffset % block_w;
 		}
 
+		if (vpstate->afbc_half_block_en)
+			block_h /= 2;
+
 		if (vpstate->ymirror_en) {
 			transform_tmp = vir_height - act_yoffset - height;
-			transform_yoffset = transform_tmp % 4;
+			transform_yoffset = transform_tmp % block_h;
 		} else {
-			transform_yoffset = act_yoffset % 4;
+			transform_yoffset = act_yoffset % block_h;
 		}
 
 		return (transform_xoffset & 0x3f) | ((transform_yoffset & 0x3f) << 16);
@@ -4183,11 +4295,44 @@
 		 */
 		if (vop2->version == VOP_VERSION_RK3588) {
 			struct vop2_power_domain *esmart_pd = vop2_find_pd_by_id(vop2, VOP2_PD_ESMART);
+			u32 pd_offset = esmart_pd->data->regs->pd.offset;
 
+			/*
+			 * Get power_ctrl default value and backup to regsbak,
+			 * so we can config pd register correctly as expected.
+			 */
+			vop2->regsbak[pd_offset >> 2] = vop2_readl(vop2, pd_offset);
 			if (vop2_power_domain_status(esmart_pd))
 				esmart_pd->on = true;
 			else
 				vop2_power_domain_on(esmart_pd);
+
+			if (vop2->data->nr_dscs) {
+				struct vop2_dsc *dsc;
+				int i = 0;
+
+				for (i = 0; i < vop2->data->nr_dscs; i++) {
+					dsc = &vop2->dscs[i];
+
+					if (!dsc->pd)
+						continue;
+
+					/* To access dsc register must after enable dsc pd and release reset */
+					if (!vop2_power_domain_status(dsc->pd) ||
+					    !VOP_MODULE_GET(vop2, dsc, rst_deassert))
+						continue;
+
+					dsc->enabled = VOP_MODULE_GET(vop2, dsc, dsc_en);
+
+					if (dsc->enabled) {
+						dsc->attach_vp_id = VOP_MODULE_GET(vop2, dsc,
+										   dsc_port_sel);
+						dsc->pd->vp_mask = BIT(dsc->attach_vp_id);
+						dsc->pd->on = true;
+						dsc->pd->ref_count++;
+					}
+				}
+			}
 		} else {
 			struct vop2_power_domain *pd, *n;
 
@@ -4453,7 +4598,7 @@
 		VOP_CTRL_SET(vop2, dma_stop, 1);
 		if (vop2->aclk_mode_rate[ROCKCHIP_VOP_ACLK_RESET_MODE] &&
 		    clk_get_rate(vop2->aclk) > vop2->aclk_mode_rate[ROCKCHIP_VOP_ACLK_RESET_MODE])
-			vop2_devfreq_set_aclk(crtc, ROCKCHIP_VOP_ACLK_RESET_MODE);
+			vop2_set_aclk_rate(crtc, ROCKCHIP_VOP_ACLK_RESET_MODE, NULL);
 		rockchip_drm_dma_detach_device(vop2->drm_dev, vop2->dev);
 		vop2->is_iommu_enabled = false;
 	}
@@ -4893,12 +5038,14 @@
 		if (dual_channel) {
 			vop2_power_domain_put(vop2->dscs[0].pd);
 			vop2_power_domain_put(vop2->dscs[1].pd);
+			vop2_wait_power_domain_off(vop2->dscs[0].pd);
+			vop2_wait_power_domain_off(vop2->dscs[1].pd);
 			vop2->dscs[0].pd->vp_mask = 0;
 			vop2->dscs[1].pd->vp_mask = 0;
 			vop2->dscs[0].attach_vp_id = -1;
 			vop2->dscs[1].attach_vp_id = -1;
 		} else {
-			vop2_power_domain_put(vop2->dscs[vcstate->dsc_id].pd);
+			vop2_power_domain_put_sync(vop2->dscs[vcstate->dsc_id].pd);
 			vop2->dscs[vcstate->dsc_id].pd->vp_mask = 0;
 			vop2->dscs[vcstate->dsc_id].attach_vp_id = -1;
 		}
@@ -7847,7 +7994,7 @@
 	.crtc_output_post_enable = vop2_crtc_output_post_enable,
 	.crtc_output_pre_disable = vop2_crtc_output_pre_disable,
 	.crtc_set_color_bar = vop2_crtc_set_color_bar,
-	.set_aclk = vop2_devfreq_set_aclk,
+	.set_aclk = vop2_set_aclk_rate,
 	.get_crc = vop2_crtc_get_crc,
 };
 
@@ -11719,12 +11866,12 @@
 			VOP_CTRL_SET(vop2, dma_stop, 1);
 		if (vop2->aclk_mode_rate[ROCKCHIP_VOP_ACLK_RESET_MODE] &&
 		    clk_get_rate(vop2->aclk) > vop2->aclk_mode_rate[ROCKCHIP_VOP_ACLK_RESET_MODE]) {
-			vop2_devfreq_set_aclk(crtc, ROCKCHIP_VOP_ACLK_RESET_MODE);
+			vop2_set_aclk_rate(crtc, ROCKCHIP_VOP_ACLK_RESET_MODE, NULL);
 			enter_vop_aclk_reset_mode = true;
 		}
 		ret = rockchip_drm_dma_attach_device(vop2->drm_dev, vop2->dev);
 		if (enter_vop_aclk_reset_mode)
-			vop2_devfreq_set_aclk(crtc, aclk_mode);
+			vop2_set_aclk_rate(crtc, aclk_mode, NULL);
 		if (ret) {
 			vop2->is_iommu_enabled = false;
 			vop2_disable_all_planes_for_crtc(crtc);
 			
