diff -uNr a/arch/arm64/boot/dts/rockchip/rk3588s-nanopi-r6-common.dtsi b/arch/arm64/boot/dts/rockchip/rk3588s-nanopi-r6-common.dtsi
--- a/arch/arm64/boot/dts/rockchip/rk3588s-nanopi-r6-common.dtsi	2024-03-11 00:45:40.000000000 +0100
+++ b/arch/arm64/boot/dts/rockchip/rk3588s-nanopi-r6-common.dtsi	2024-03-11 12:24:59.171687463 +0100
@@ -146,8 +146,6 @@
 &gpu {
 	mali-supply = <&vdd_gpu_s0>;
 	mem-supply = <&vdd_gpu_mem_s0>;
-	upthreshold = <60>;
-	downdifferential = <30>;
 	status = "okay";
 };


