From 542e92dd79085ff6f033a00b755a9a6e5539b6ea Mon Sep 17 00:00:00 2001
From: Seinist <phobossbk@googlemail.com>
Date: Mon, 7 Mar 2022 06:18:56 +0100
Subject: [PATCH] Do not present CEC toasts unless there is an error or library

---
 xbmc/peripherals/devices/PeripheralCecAdapter.cpp | 14 --------------
 1 file changed, 14 deletions(-)

diff --git a/xbmc/peripherals/devices/PeripheralCecAdapter.cpp b/xbmc/peripherals/devices/PeripheralCecAdapter.cpp
index 9bc0979d6b9e3..3a1b23b48ef2f 100644
--- a/xbmc/peripherals/devices/PeripheralCecAdapter.cpp
+++ b/xbmc/peripherals/devices/PeripheralCecAdapter.cpp
@@ -308,13 +308,6 @@ bool CPeripheralCecAdapter::OpenConnection(void)
   CLog::Log(LOGDEBUG, "{} - opening a connection to the CEC adapter: {}", __FUNCTION__,
             m_strComPort);
 
-  // scanning the CEC bus takes about 5 seconds, so display a notification to inform users that
-  // we're busy
-  std::string strMessage =
-      StringUtils::Format(g_localizeStrings.Get(21336), g_localizeStrings.Get(36000));
-  CGUIDialogKaiToast::QueueNotification(CGUIDialogKaiToast::Info, g_localizeStrings.Get(36000),
-                                        strMessage);
-
   bool bConnectionFailedDisplayed(false);
 
   while (!m_bStop && !bIsOpen)
@@ -1639,10 +1632,6 @@ bool CPeripheralCecAdapterUpdateThread::SetInitialConfiguration(void)
 
   m_adapter->m_bIsReady = true;
 
-  // and let the gui know that we're done
-  CGUIDialogKaiToast::QueueNotification(CGUIDialogKaiToast::Info, g_localizeStrings.Get(36000),
-                                        strNotification);
-
   std::unique_lock<CCriticalSection> lock(m_critSection);
   m_bIsUpdating = false;
   return true;
@@ -1688,9 +1677,6 @@ void CPeripheralCecAdapterUpdateThread::Process(void)
         UpdateAudioSystemStatus();
       }
 
-      CGUIDialogKaiToast::QueueNotification(CGUIDialogKaiToast::Info, g_localizeStrings.Get(36000),
-                                            g_localizeStrings.Get(bConfigSet ? 36023 : 36024));
-
       {
         std::unique_lock<CCriticalSection> lock(m_critSection);
         if ((bUpdate = m_bNextConfigurationScheduled) == true)
