From 61c5dd3115b32a5c5369c334493ea5d196e3cb2b Mon Sep 17 00:00:00 2001
From: afl1 <afl2001@gmail.com>
Date: Fri, 31 May 2019 13:38:16 +0200
Subject: [PATCH] VideoPlayer: increase delay for live TV playback start

---
 xbmc/cores/VideoPlayer/VideoPlayer.cpp | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/xbmc/cores/VideoPlayer/VideoPlayer.cpp b/xbmc/cores/VideoPlayer/VideoPlayer.cpp
index 24e0899260..5685105067 100644
--- a/xbmc/cores/VideoPlayer/VideoPlayer.cpp
+++ b/xbmc/cores/VideoPlayer/VideoPlayer.cpp
@@ -2044,7 +2044,7 @@ void CVideoPlayer::HandlePlaySpeed()
       else if (m_CurrentAudio.starttime != DVD_NOPTS_VALUE && m_CurrentAudio.packets > 0)
       {
         if (m_pInputStream->IsRealtime())
-          clock = m_CurrentAudio.starttime - m_CurrentAudio.cachetime - DVD_MSEC_TO_TIME(800);
+          clock = m_CurrentAudio.starttime - m_CurrentAudio.cachetime - DVD_MSEC_TO_TIME(1000);
         else
           clock = m_CurrentAudio.starttime - m_CurrentAudio.cachetime;
 

